### PR TITLE
Fix bugged pet sit button and taunt.

### PR DESCRIPTION
**Pull request by noudess**
The pet window sit button is often out of sync.  This PR fixes this issue.  Also pet's were frequently still marked as sitting internal to the server due to the owner being attacked, or other issues.  This is because the anim state was not being correctly updated.

I tested the behavior on live and made it match.  A pet on guard will always return to guard after a battle.  A pet sitting, will switch to follow mode after a battle.  This fix is here as well.

After 2 weeks of testing, I no longer see the pet sit button out of sync, and no longer see (harder to see, but can be done with a char in his teens) pet sitting regen while pet is actually standing.

After hearing from @Trust that the taunt button was also out of sync often, I looked into that as well.

The client stores the status of an active pet's taunt setting.  This is stored across sessions.  The old approach was to try and force taunt on, on pet creation as well as logging on and zoning.   Client messages, sometimes in response to out attempts, would end up toggling our taunt value away from what we just set it to.

Taunt now behaves as follows:

**Clients previous to UF**

Only client on/off clicks change state.  Default is taunt on.  Zoning and logging on both default to true.

UF+ Clients:

**On 1st time pet summoning via spell:**

The pet is created with taunting off.  This matches the client.

**On logging in with an existing pet:**

The client sends a message if the pet was taunting when you logged off, which will turn it on, since pet creation now defaults to  taunting=false in the constructor.

**On zoning:**

The pet's state is set from it's saved value in the database.  The client is not involved, but will match as pet state is preserved by the client across zoning.

**On suspend minion:**

The pet's state is set from it's saved value in the database.  The client is not involved, but will match as pet state is preserved by the client across zoning.

NOTE: After this update, existing pets might be out of sync, because I have no way of knowing if your client has the button depressed or not.  My guess is that they will work.  I do not use the db saved version at all on login from character_select, so the 1st login should sync with the client.

### CODE REVIEW

**Comment by noudess** [comment]
At the behest of @Trust I am currently adding fixes for the taunt button along the same lines.

**Comment by freyonsecurity** [comment]
I played with this patch for 6 hours nonstop without any problems as a pet magician. Very good improvement!

File: zone/attack.cpp [hunk 1]
```diff
@@ -3445,6 +3445,22 @@ void Mob::CommonDamage(Mob* attacker, int &damage, const uint16 spell_id, const
...
+			// If a client pet is damaged while sitting, stand, fix sit button,
+			// and remove sitting regen.  Removes bug where client clicks sit
+			// during battle and gains pet hp-regen and bugs the sit button.
+			if (IsPet()) {
+				Mob *owner = this->GetOwner();
+				if (owner && owner->IsClient()) {
+					if (GetPetOrder() == SPO_Sit) {
+						SetPetOrder(SPO_Follow);
+					}
+				// fix GUI sit button to be unpressed and stop sitting regen
```
Comment (mackal) [inline zone/attack.cpp:3458]: wrong indention level

**Review by mackal — CHANGES_REQUESTED** [review]

File: zone/pets.cpp [hunk 2]
```diff
@@ -433,7 +433,21 @@ Pet::Pet(NPCType *type_data, Mob *owner, PetType type, uint16 spell_id, int16 po
...
+	// All pets start at false on newer clients. The client
+	// turns it on and tracks the state.
+	taunting=false;
+
+	// Older clients didn't track state, and default taunting is on (per @mackal)
+	// Familiar and animation pets don't get taunt until an AA.
+	if (owner && owner->IsClient()) {
+		if (!(owner->CastToClient()->ClientVersionBit() & EQ::versions::maskUFAndLater)) {
+			if ((typeofpet != petFamiliar && typeofpet != petAnimation) || 
+				GetAA(aaAnimationEmpathy) >= 3) {
```
[the remark this answers is not in the document]
Comment (noudess) [reply zone/pets.cpp:446]: Can you give me how you would code it?  Just replace GetAA or that entire if?

**Review by mackal — APPROVED** [review]

**Review by Akkadius — APPROVED** [review]
Looks great, good work!

**Pull request merged**